Nombre de estudiante:

**Objetivo**

Implementar los método clásicos iterativos de Jacobi y Gauss Seidel

**Métodos iterativos clásicos**

Para garantizar la convergencia de un método iterativo estacionario general, el radio espectral de la matriz de iteración debe ser menor
que $1$. Dado que cada uno de los métodos iterativos clásicos está caracterizado por la matriz de iteración, entonces debemos verificar que éstas satisfacen la condición: $\rho(B)<1$ donde $B$ es la matriz de iteración que caracteriza al método.

**Convergencia del Método de Jacobi**

El método de Jacobi es convergente para cualquier valor inicial si $A$ es una matriz diagonal dominante (por filas o columnas).  
El valor absoluto del elemento diagonal debe ser mayor o igual que la suma de los demás elementos de esa fila.

# Métodos Iterativos: Jacobi y Gauss-Seidel

Queremos resolver el sistema lineal:

$$
A\mathbf{x}=\mathbf{b}
$$

La idea de los métodos iterativos es escribir el sistema en la forma:

$$
\mathbf{x}=B\mathbf{x}+\mathbf{d}
$$

donde:

- $B$ = matriz de iteración
- $\mathbf{d}$ = vector constante

Partiendo de una aproximación inicial:

$$
\mathbf{x}^{(1)}
$$

se generan nuevas aproximaciones usando:

$$
\mathbf{x}^{(k+1)}=
B\mathbf{x}^{(k)}+\mathbf{d}
$$

para:

$$
k=1,2,3,\dots
$$

Esto produce una sucesión:

$$
\mathbf{x}^{(1)},
\mathbf{x}^{(2)},
\mathbf{x}^{(3)},
\dots
$$

que idealmente converge a la solución exacta:

$$
\mathbf{x}
$$

---

## Idea general

Los métodos iterativos responden dos preguntas:

### (a) ¿Cómo escribir?

$$
A\mathbf{x}=b
$$

en la forma:

$$
\mathbf{x}=B\mathbf{x}+d
$$

### (b) ¿Cuándo converge?

Es decir:

- ¿Qué aproximación inicial usar?
- ¿Bajo qué condiciones converge?

---

## Métodos estudiados

Existen tres métodos iterativos clásicos:

- Jacobi
- Gauss-Seidel
- SOR (Successive Over Relaxation)

La diferencia principal entre ellos es:

> **Cómo construyen la matriz de iteración $B$ y el vector $d$.**

En este laboratorio implementaremos:

### Jacobi

y

### Gauss-Seidel

para comparar velocidad de convergencia y comportamiento.

# Fórmula general del método de Jacobi

Partiendo del sistema:

$$
A\mathbf{x}=\mathbf{b}
$$

cada ecuación:

$$
a_{i1}x_1+a_{i2}x_2+\cdots+a_{ii}x_i+\cdots+a_{in}x_n=b_i
$$

se despeja respecto a:

$$
x_i
$$

obteniéndose:

$$
x_i^{(k+1)}
=
\frac{
b_i-
\sum_{j\neq i}^{n}
a_{ij}x_j^{(k)}
}{
a_{ii}
}
\qquad i=1,2,\dots,n
$$

Observación:

> Jacobi utiliza únicamente valores de la iteración anterior.

---

# Fórmula general del método de Gauss-Seidel

El método de Gauss-Seidel modifica Jacobi reutilizando inmediatamente los nuevos valores calculados.

La fórmula general es:

$$
x_i^{(k+1)}
=
\frac{
1
}{
a_{ii}
}
\left(
b_i
-
\sum_{j=1}^{i-1}
a_{ij}x_j^{(k+1)}
-
\sum_{j=i+1}^{n}
a_{ij}x_j^{(k)}
\right)
$$

para:

$$
i=1,2,\dots,n
$$

Observación:

> Gauss-Seidel utiliza inmediatamente los valores recién calculados dentro de la misma iteración.

---

## Diferencia principal

Jacobi:

$$
x^{(k)}
\rightarrow
x^{(k+1)}
$$

Gauss-Seidel:

$$
x_1^{(k+1)}
\rightarrow
x_2^{(k+1)}
\rightarrow
x_3^{(k+1)}
\rightarrow
\cdots
$$

Por esta razón:

> Gauss-Seidel normalmente converge en menos iteraciones.

In [4]:
import numpy as np                                        # Librería numérica

def jacobi(A,b,x0,n_iter):                               # Método de Jacobi

    n = len(b)                                            # Número de incógnitas

    x = x0.copy()                                         # Aproximación inicial

    historial = []                                        # Guarda iteraciones

    for k in range(n_iter):                               # Ciclo principal

        x_new = np.zeros_like(x)                          # Nuevo vector iterado

        for i in range(n):                                # Recorre filas

            suma = 0                                      # Inicializa sumatoria

            for j in range(n):                            # Recorre columnas

                if j != i:                                # Ignora diagonal

                    #------------------------------------ # Σ aij*xj iteración anterior

            #---------------------------------------------# Fórmula de Jacobi

        historial.append(x_new.copy())                    # Guarda iteración

        #-------------------------------------------------# Norma infinito error

        print("iter",k+1,
              "x =",np.round(x_new,4),
              "error =",round(error,8))                  # Imprime iteración

        x = x_new.copy()                                  # Actualiza iteración

    return historial                                      # Devuelve historial


# --------------------------------------------------
# EJEMPLO DEL LIBRO BIRSWA
# --------------------------------------------------

A = np.array([[5,1,1],                                    # Matriz A
              [1,5,1],
              [1,1,5]],dtype=float)

b = np.array([7,7,7],dtype=float)                         # Vector b

x0 = np.array([0,0,0],dtype=float)                        # Aproximación inicial

J = jacobi(A,b,x0,n_iter=7)                               # Ejecuta Jacobi

print()

print("Última aproximación:")

print(np.round(J[-1],6))                                  # Última iteración

iter 1 x = [1.4 1.4 1.4] error = 1.4
iter 2 x = [0.84 0.84 0.84] error = 0.56
iter 3 x = [1.064 1.064 1.064] error = 0.224
iter 4 x = [0.9744 0.9744 0.9744] error = 0.0896
iter 5 x = [1.0102 1.0102 1.0102] error = 0.03584
iter 6 x = [0.9959 0.9959 0.9959] error = 0.014336
iter 7 x = [1.0016 1.0016 1.0016] error = 0.0057344

Última aproximación:
[1.001638 1.001638 1.001638]


In [5]:
import numpy as np                                      # Librería numérica

# COMPLETAR EL CLASE EL CÓDIGO

# --------------------------------------------------
# EJEMPLO DEL LIBRO
# --------------------------------------------------

A = np.array([[5,1,1],                                  # Matriz A
              [1,5,1],
              [1,1,5]],dtype=float)

b = np.array([7,7,7],dtype=float)                       # Vector b

x0 = np.array([0,0,0],dtype=float)                      # Aproximación inicial

GS = gauss_seidel(A,b,x0,n_iter=7)                      # Ejecuta GS

print()

print("Última aproximación:")

print(np.round(GS[-1],6))                               # Última iteración

iter 1 x = [1.4   1.12  0.896] error = 1.4
iter 2 x = [0.9968 1.0214 0.9964] error = 0.4032
iter 3 x = [0.9964 1.0014 1.0004] error = 0.01999872
iter 4 x = [0.9996 1.     1.0001] error = 0.00318546
iter 5 x = [1. 1. 1.] error = 0.00035964
iter 6 x = [1. 1. 1.] error = 1.48e-05
iter 7 x = [1. 1. 1.] error = 1.29e-06

Última aproximación:
[1. 1. 1.]


In [6]:
# Ejecuta ambos métodos (ya implementados previamente)

#------------------------------------------------          # Historial Jacobi

#-----------------------------------------------           # Historial Gauss-Seidel


print("-"*80)

print(f"{'k':<5}{'Jacobi':<35}{'Gauss-Seidel'}")

print("-"*80)

#completar el clase
#-------------------
#-------------------

    print(f"{k+1:<5}{str(xj):<35}{str(xgs)}")

print("-"*80)

iter 1 x = [1.4 1.4 1.4] error = 1.4
iter 2 x = [0.84 0.84 0.84] error = 0.56
iter 3 x = [1.064 1.064 1.064] error = 0.224
iter 4 x = [0.9744 0.9744 0.9744] error = 0.0896
iter 5 x = [1.0102 1.0102 1.0102] error = 0.03584
iter 6 x = [0.9959 0.9959 0.9959] error = 0.014336
iter 7 x = [1.0016 1.0016 1.0016] error = 0.0057344
iter 1 x = [1.4   1.12  0.896] error = 1.4
iter 2 x = [0.9968 1.0214 0.9964] error = 0.4032
iter 3 x = [0.9964 1.0014 1.0004] error = 0.01999872
iter 4 x = [0.9996 1.     1.0001] error = 0.00318546
iter 5 x = [1. 1. 1.] error = 0.00035964
iter 6 x = [1. 1. 1.] error = 1.48e-05
iter 7 x = [1. 1. 1.] error = 1.29e-06
--------------------------------------------------------------------------------
k    Jacobi                             Gauss-Seidel
--------------------------------------------------------------------------------
1    [1.4 1.4 1.4]                      [1.4   1.12  0.896]
2    [0.84 0.84 0.84]                   [0.9968 1.0214 0.9964]
3    [1.064 1.064 1.

# Conclusiones: Rapidez y Convergencia

## Rapidez de convergencia

Observamos que:

- **Jacobi necesita más iteraciones** para aproximarse a la solución.
- **Gauss-Seidel converge más rápidamente.**

### ¿Por qué ocurre esto?

### Jacobi

En Jacobi, toda la nueva iteración se calcula usando únicamente los valores anteriores:

$$
x^{(k+1)}=Bx^{(k)}+d
$$

Es decir:

> **espera terminar toda la iteración antes de utilizar nuevos valores.**

Por ejemplo:

Si estamos calculando:

$$
x_1,x_2,x_3
$$

todos utilizan información de:

$$
x^{(k)}
$$

---

### Gauss-Seidel

Gauss-Seidel funciona diferente.

Apenas calcula un nuevo valor:

$$
x_1^{(k+1)}
$$

lo utiliza inmediatamente para calcular:

$$
x_2^{(k+1)}
$$

y luego:

$$
x_3^{(k+1)}
$$

Es decir:

> **usa información más reciente dentro de la misma iteración.**

Visualmente:

Jacobi:

$$
(x^{(k)})\rightarrow(x^{(k+1)})
$$

Gauss-Seidel:

$$
x_1^{(k+1)}
\rightarrow
x_2^{(k+1)}
\rightarrow
x_3^{(k+1)}
$$

Por esta razón normalmente necesita menos iteraciones.

---

## Comparación observada

| Iteración | Jacobi | Gauss-Seidel |
|---|---:|---:|
| 1 | 1.4000 | 1.4000 |
| 4 | 0.9744 | 0.9996 |
| 6 | 0.9959 | 1.0000 |

Observamos que:

> **Gauss-Seidel alcanza la solución aproximada más rápidamente.**

---

## Convergencia

La matriz utilizada fue:

$$
A=
\begin{bmatrix}
5&1&1\\
1&5&1\\
1&1&5
\end{bmatrix}
$$

Esta matriz es estrictamente diagonal dominante por filas porque:

$$
5>1+1
$$

Por teoría:

> **Si la matriz es estrictamente diagonal dominante por filas, Jacobi y Gauss-Seidel convergen.**

---

## Criterio de parada utilizado

Se utilizó:

$$
\|x^{(k+1)}-x^{(k)}\|_{\infty}<tol
$$

donde:

$$
\|v\|_\infty=
\max_i |v_i|
$$

Esto significa:

> **Se detiene cuando el cambio máximo entre dos iteraciones consecutivas es suficientemente pequeño.**

---

## Conclusión final

> Ambos métodos convergen debido a que la matriz es estrictamente diagonal dominante por filas. Sin embargo, Gauss-Seidel converge más rápidamente porque reutiliza inmediatamente los valores recién calculados dentro de cada iteración.

# Ejercicios: Métodos de Jacobi y Gauss-Seidel

Usar:

- Aproximación inicial:

$$
x^{(0)}=(0,0,\dots,0)^T
$$

- Tolerancia:

$$
10^{-6}
$$

---

# Ejercicio 1 (4×4)

Resolver:

$$
\begin{bmatrix}
12&1&1&1\\
2&11&1&1\\
1&2&10&1\\
1&1&2&9
\end{bmatrix}
\begin{bmatrix}
x_1\\x_2\\x_3\\x_4
\end{bmatrix}
=
\begin{bmatrix}
20\\13\\13\\10
\end{bmatrix}
$$

Respuesta:

$$
x=(1,0,1,1)^T
$$

---

# Ejercicio 2 (4×4)

Resolver:

$$
\begin{bmatrix}
10&1&1&1\\
1&9&1&1\\
1&1&8&1\\
1&1&1&7
\end{bmatrix}
\begin{bmatrix}
x_1\\x_2\\x_3\\x_4
\end{bmatrix}
=
\begin{bmatrix}
14\\15\\15\\13
\end{bmatrix}
$$

Respuesta:

$$
x=(1,1,1,1.5)^T
$$

---

# Ejercicio 3 (5×5)

Resolver:

$$
\begin{bmatrix}
15&1&1&1&1\\
1&14&1&1&1\\
1&1&13&1&1\\
1&1&1&12&1\\
1&1&1&1&11
\end{bmatrix}
\begin{bmatrix}
x_1\\x_2\\x_3\\x_4\\x_5
\end{bmatrix}
=
\begin{bmatrix}
20\\18\\16\\17\\14
\end{bmatrix}
$$

Respuesta:

$$
x=(1,0.5,0.5,1,1)^T
$$

---

# Ejercicio 4 (5×5)

Resolver:

$$
\begin{bmatrix}
20&2&1&1&1\\
1&18&2&1&1\\
1&1&17&2&1\\
1&1&1&16&2\\
1&1&1&1&15
\end{bmatrix}
\begin{bmatrix}
x_1\\x_2\\x_3\\x_4\\x_5
\end{bmatrix}
=
\begin{bmatrix}
27\\26\\23\\22\\17
\end{bmatrix}
$$

Respuesta:

$$
x=(1,1,0.5,1,0.8)^T
$$

---

# Ejercicio 5 (5×5)

Resolver:

$$
\begin{bmatrix}
18&1&1&1&1\\
2&17&1&1&1\\
2&2&16&1&1\\
2&2&2&15&1\\
2&2&2&2&14
\end{bmatrix}
\begin{bmatrix}
x_1\\x_2\\x_3\\x_4\\x_5
\end{bmatrix}
=
\begin{bmatrix}
24\\20\\18\\16\\14
\end{bmatrix}
$$

Respuesta aproximada:

$$
x=(1,0.5,0.5,0.5,0.5)^T
$$

---

## Preguntas

1. ¿Cuál método converge en menos iteraciones?

2. Compare el error obtenido en cada método.

3. ¿Qué sucede si cambia la aproximación inicial?

4. ¿Qué ocurre si la matriz deja de ser diagonal dominante?

5. Verifique que:

$$
Ax\approx b
$$

con la solución encontrada.